File: Bughound/employees.py
import cherrypy
import json
import cx_Oracle
from datetime import datetime
import xml.etree.ElementTree
import os

import sys
sys.path.insert(0, 'c:\inetpub\wwwroot\CSULB\CECS544\ReusablePython')
from DB_Connection import db_CECS544

class employees:
    # allow the cherrypy server to serve these files
    exposed = True
    @cherrypy.tools.accept(media='text/html')


    def __init__(self):
        self.dispatch = {
            'Search' : self.search_employees,
            'Add': self.add_employee,
            'Delete': self.delete_employee,
            'Populate' : self.populate_employees,
            'Update' : self.update_employee,
            'ASCII' : self.ExportEmployeeData_ASCII,
            'XML' : self.ExportEmployeeData_XML
        }

    def POST(self, params):
        # create a connection to the database
        try:
            self.CECS544_DB = db_CECS544()
            cur = self.CECS544_DB.conn.cursor()
        except:
            raise ValueError(err)

        # data structure that we will use to return values back to ajax call
        self.sendData = {'Result': '', 'Data': [], 'Error': ''}

        # load the params into a python dictionary so we can use them
        self.Params = json.loads(params)

        try:
            # call the method specified in the AJAX call
            self.dispatch[self.Params['Method']](cur)

            return json.dumps(self.sendData)
        except:
            return json.dumps(self.sendData)

    def add_employee(self, cur):
        sql = '''
        INSERT INTO EMPLOYEE
            (emp_name, emp_username, emp_password, emp_role)
        VALUES
            (:Employee_Name, :Employee_Username, :Employee_Password, :Employee_Role)
        '''

        try:
            cur.execute(sql, Employee_Name=self.Params['Employee_Name'], Employee_Username=self.Params['Employee_Username'], Employee_Password=self.Params['Employee_Password'], Employee_Role=self.Params['Employee_Role'])
            self.CECS544_DB.conn.commit()

        except cx_Oracle.IntegrityError as e:
            self.sendData['Result']='PK Violation'
            raise ValueError()

        self.sendData['Result']='Success'

    def update_employee(self, cur):
        sql = '''
        UPDATE EMPLOYEE
            SET
            emp_username = :username,
            emp_name = :name,
            emp_password = :password,
            emp_role = :role
        WHERE
            emp_id = :id
        '''

        cur.execute(sql, username=self.Params['emp_username'], name=self.Params['emp_name'], password=self.Params['emp_password'], role=self.Params['emp_role'], id=self.Params['emp_id'])
        self.CECS544_DB.conn.commit()

        self.sendData['Result'] = 'Success'

    def delete_employee(self, cur):
        sql = '''
        DELETE FROM EMPLOYEE
        WHERE
            emp_id = :id
        '''

        cur.execute(sql, id=self.Params['EmpID'])
        self.CECS544_DB.conn.commit()

        self.sendData['Result'] = 'Success'

    def search_employees(self, cur):
        sql = '''
        select EMP_ID, EMP_USERNAME, EMP_NAME, EMP_ROLE
        from EMPLOYEE
        WHERE 1=1
        '''
        bindVars = {}
        if self.Params['Employee_ID'] != "":
            sql = sql + ' AND EMP_ID = :Employee_ID'
            bindVars['Employee_ID'] = self.Params['Employee_ID']

        if self.Params['Employee_Username'] != "":
            sql = sql + ' AND EMP_USERNAME = :Employee_Username'
            bindVars['Employee_Username'] = self.Params['Employee_Username']

        if self.Params['Employee_Name'] != "":
            sql = sql + ' AND EMP_NAME = :Employee_Name'
            bindVars['Employee_Name'] = self.Params['Employee_Name']

        if self.Params['Employee_Role'] != "PleaseSelect":
            sql = sql + ' AND EMP_ROLE = :Employee_Role'
            bindVars['Employee_Role'] = self.Params['Employee_Role']

        cur.execute(sql, bindVars) # execute the sql statement
        allRows = cur.fetchall() # get the results on the query

        # loop through the query results
        for row in allRows:
            # save the data to our strucutre we are sending back via AJAX
            self.sendData['Data'].append({
                'ID': row[0],
                'Username': row[1],
                'Name': row[2],
                'Role': row[3]
            })

        self.sendData['Result'] = 'Success'

    def populate_employees(self, cur):
        sql = '''
        select EMP_ID, EMP_USERNAME, EMP_NAME, EMP_ROLE, EMP_PASSWORD
        from EMPLOYEE
        WHERE EMP_ID=:emp_id
        '''

        cur.execute(sql,emp_id=self.Params['emp_id'])
        allRows = cur.fetchall()

        for row in allRows:
            self.sendData['Data'].append({
                'ID': row[0],
                'Username' : row[1],
                'Name' : row[2],
                'Role' : row[3],
                'Password' : row[4]
            })

        self.sendData['Result'] = 'Success'

    def ExportEmployeeData_ASCII(self, cur):
        sql = '''
        SELECT EMP_ID, EMP_NAME, EMP_USERNAME, EMP_PASSWORD,  EMP_ROLE
        FROM EMPLOYEE
        '''

        cur.execute(sql)
        allRows = cur.fetchall()

        asciiExport = 'EMP_ID, EMP_NAME, EMP_USERNAME, EMP_PASSWORD,  EMP_ROLE'
        for row in allRows:
            asciiExport = asciiExport + '\n%s\t%s\t%s\t%s\t%s' % (row[0], row[1], row[2], row[3], row[4])

        os.chdir('c:\inetpub\wwwroot\CSULB\CECS544\Bughound\Export')
        file = open("EmployeeExport_ASCII.txt", "w")
        file.write(asciiExport)
        file.close()

        self.sendData['Result'] = 'Success'
        self.sendData['FileName'] = 'Export\EmployeeExport_ASCII.txt'

    def ExportEmployeeData_XML(self, cur):
        sql = '''
        select dbms_xmlgen.getxml('select * from EMPLOYEE') xml from dual
        '''

        cur.execute(sql)
        xml_data = str(cur.fetchone()[0])

        os.chdir('c:\inetpub\wwwroot\CSULB\CECS544\Bughound\Export')
        with open("EmployeeExport_XML.xml", "w") as file:
            file.write(xml_data)

        # add timestamp as attribute of root
        et = xml.etree.ElementTree.parse('EmployeeExport_XML.xml')
        root = et.getroot()
        root.tag = "EMPLOYEES"
        root.set("timestamp", str(datetime.now()))

        for element in root.iter("ROW"):
            element.tag = "EMPLOYEE"

        # write back to file
        et.write('EmployeeExport_XML.xml')

        self.sendData['Result'] = 'Success'
        self.sendData['FileName'] = 'Export\EmployeeExport_XML.xml'

File: Bughound/Employees.js
function showEmployeesSection(){
  // show/hide the sections we want the user to see
  $('#bugs').hide();
  $('#functionalAreas').hide();
  $('#employees').show();
  $('#programs').hide();
  $('#export').hide();


  // change the active flag on the navigation bar
  $('#Home').removeClass('active');
  $('#Bugs').removeClass('active');
  $('#FunctionalArea').removeClass('active');
  $('#Employees').addClass('active');
  $('#Programs').removeClass('active');
  $('#Export').removeClass('active');
}

function ResetEmpSearch(){
    $('#emp_id').val('');
    $('#emp_username').val('');
    $('#emp_name').val('');
    $('#emp_role').val('PleaseSelect');
}

function ResetEmpAdd(){
    $('#add_emp_id').val('');
    $('#add_emp_username').val('');
    $('#add_emp_password').val('');
    $('#add_emp_name').val('');
    $('#add_emp_role').val('PleaseSelect');
}

function add_employee(){
    var params = {
        'Method' : 'Add',
        'Employee_Name' : $('#add_emp_name').val(),
        'Employee_Username' : $('#add_emp_username').val(),
        'Employee_Password' : $('#add_emp_password').val(),
        'Employee_Role' : $('#add_emp_role').val()
    }

    if (params['Employee_Name']!='' && params['Employee_Username']!='' && params['Employee_Password']!='' && params['Employee_Role']!='PleaseSelect'){
         // data needs to be formatted before it can be sent via ajax
        var formData = new FormData()
        formData.append('params', JSON.stringify(params));

        // make the ajax call
        var AJAX_Call = getData(formData, './employees');

        // wait for the ajax call to finish then execute...
        $.when(AJAX_Call).then(function (AJAX_Response){
            if (AJAX_Response['Result'] == 'Success'){
                // clear the values the user entered
                $('#add_emp_name').val('');
                $('#add_emp_username').val('');
                $('#add_emp_password').val('');
                $('#add_emp_role').val('');

                // let the user know it was successful
                alert('You have successfully added a new employee');
            }
            else if (AJAX_Response['Result']=='PK Violation')
                alert('This record already exists. Add Failed.');
        })

        // the AJAX call was not issued succesfully
        .fail(function(load){
            alert("The webpage is unable to load, please contact the system admin")
        })
    }

    else
        alert("Please fill out all required fields");
}

function search_employees(){
    var params = {
        'Method' : 'Search',
        'Employee_ID' : $('#emp_id').val(),
        'Employee_Username' : $('#emp_username').val(),
        'Employee_Name' : $('#emp_name').val(),
        'Employee_Role' : $('#emp_role').val()
    }

    // data needs to be formatted before it can be sent via ajax
    var formData = new FormData()
    formData.append('params', JSON.stringify(params));

    // make the ajax call
    var AJAX_Call = getData(formData, './employees');

    // wait for the ajax call to finish then execute...
    $.when(AJAX_Call).then(function (AJAX_Response){
        if (AJAX_Response['Result'] == 'Success') {
            // clear any existing search results
            $('#employeeSearchResults').empty();
            if (AJAX_Response['Data'].length > 0){
                //  creeate a table to dsiplay the results
                $('#employeeSearchResults').append('<table id="EmployeeResultsTable" class="ResultsTable">');

                // add a header row to the table
                $('#EmployeeResultsTable').append('<thead><th>Employee ID</th><th>Employee Username</th><th>Employee Name</th><th>Employee Role</th></thead>');

                // loop through the search results and add them to the results table
                var tr;
                for (var i = 0; i < AJAX_Response['Data'].length; i++) {
                    tr = $('<tr/>'); // this is jquery short hand for adding a new row object
                    tr.append('<td onclick="open_employee(\'' + AJAX_Response['Data'][i].ID + '\')" class="link">' + AJAX_Response['Data'][i].ID + '</td>'); // populate the new row, cell by cell
                    tr.append('<td>' + AJAX_Response['Data'][i].Username + '</td>'); // populate the new row, cell by cell
                    tr.append('<td>' + AJAX_Response['Data'][i].Name + '</td>'); // populate the new row, cell by cell
                    tr.append('<td>' + AJAX_Response['Data'][i].Role + '</td>'); // populate the new row, cell by cell
                    $('#EmployeeResultsTable').append(tr); // add the row you just built to the table
                }
            }
            else
            $('#employeeSearchResults').append('No results found');
        }
    })

    // the AJAX call was not issued succesfully
	.fail(function(load){
		alert("The webpage is unable to load, please contact the system admin")
	})
}

function delete_employees(emp_id){
    //promt user to confirm deletion
    if (confirm("Are you sure you want to delete " + emp_id + "?")){
        var params = {
            'Method' : 'Delete',
            'EmpID' : emp_id
        }
        // data needs to be formatted before it can be sent via ajax
        var formData = new FormData()
        formData.append('params', JSON.stringify(params));

        // make the ajax call
        var AJAX_Call = getData(formData, './employees');

        // wait for the ajax call to finish then execute...
        $.when(AJAX_Call).then(function (AJAX_Response){
            if (AJAX_Response['Result'] == 'Success'){
                // re-run the search to show the user its been deleted
                search_employees();

                // let the user know it was successful
                alert('You have successfully deleted ' + emp_id);
            }
        })

        // the AJAX call was not issued succesfully
        .fail(function(load){
            alert("The webpage is unable to load, please contact the system admin");
        })
    }
}


function open_employee(emp_id){
	// set a cookie so the next page can read the cookie and know which bug to open
	setCookie('emp_id', emp_id, .5);

	// redirect to the bug editor
	window.open('./Employee_Editor.html');
}

function load_employees(){
    if (document.cookie.indexOf("EmployeeName")>=0){
        var EmployeeName = getCookie("EmployeeName");
        
        if (document.cookie.indexOf("emp_id")>=0){

            // save the empid to a variable that we will send to the search function
            var emp_id = getCookie("emp_id");

            // delete the cookie so if a user opens another page the bugID variable is reset
            setCookie('emp_id', emp_id, 0);

            populate_emp_editor(emp_id);
        }
    }
    else
		window.location.replace('http://localhost:8081/ReusableJavascript/Login.html');
}

function populate_emp_editor(emp_id){
    var params = {
        'Method': 'Populate',
        'emp_id': emp_id
    }

    // data needs to be formatted before it can be sent via ajax
    var formData = new FormData()
    formData.append('params', JSON.stringify(params));

    // make the ajax call
    var AJAX_Call = getData(formData, './employees');

    // wait for the ajax call to finish then execute...
    $.when(AJAX_Call).then(function (AJAX_Response){
        if (AJAX_Response['Result'] == 'Success') {
            // populate the form with the values that are currently in the database
            // we can hardcode [0] because only 1 row will ever be returned since we are searching
            // by the bug ID.  We could change our sendData variable in python but this way
            // everything stays consistent
            $('#emp_id').val(AJAX_Response['Data'][0]['ID']);
            $('#emp_username').val(AJAX_Response['Data'][0]['Username']);
            $('#emp_name').val(AJAX_Response['Data'][0]['Name']);
            $('#emp_password').val(AJAX_Response['Data'][0]['Password']);
            $('#emp_role').val(AJAX_Response['Data'][0]['Role']);
        }
    })

    // the AJAX call was not issued succesfully
	.fail(function(load){
		alert("The webpage is unable to load, please contact the system admin")
	})
}

function update_employee(){
    var params = {
        'Method' : 'Update',
        'emp_id' : $('#emp_id').val(),
        'emp_username' : $('#emp_username').val(),
        'emp_name' : $('#emp_name').val(),
        'emp_password' : $('#emp_password').val(),
        'emp_role' : $('#emp_role').val()
    }

    // data needs to be formatted before it can be sent via ajax
    var formData = new FormData()
    formData.append('params', JSON.stringify(params));

    // make the ajax call
    var AJAX_Call = getData(formData, './employees');

    // wait for the ajax call to finish then execute...
    $.when(AJAX_Call).then(function (AJAX_Response){
        if (AJAX_Response['Result'] == 'Success'){
            // let the user know it was successful
            alert('You have successfully saved your changes');
            window.close();
        }
    })

    // the AJAX call was not issued succesfully
	.fail(function(load){
		alert("The webpage is unable to load, please contact the system admin")
	})
}
